### PR TITLE
REVIEWED : IC Token runtime validation with hash-check and jti uniqueness

### PR DESCRIPTION
**Pull request by alchezar**
## Summary

- Runtime endpoints (handshake, refresh, report, provider-key) now validate IC Token
hash against database on every request, ensuring revoked/rotated tokens are rejected
- Added `jti` (JWT ID) claim to IcTokenClaims — UUID v4 guarantees unique JWT even
when generated in the same second with identical parameters
- Unified all auth failures to return 401 "Invalid IC Token" (prevents enumeration)

## Problem

After `regenerate` or `revoke`, old IC Tokens remained usable because runtime
endpoints only verified JWT signature — never checked if the token was still
the active one stored in `agents.ic_token_hash`.

## Changes

1. **Shared validator** (`ic_token.rs:validate_ic_token_runtime`) — JWT verification +
SHA-256 hash-check against DB in a single function
2. **jti claim** — `Uuid::new_v4()` in every `IcTokenClaims::new()`, so regenerate
in the same second produces a different JWT with a different hash
3. **Test updates** — all test helpers store `ic_token_hash`, new tests cover
regenerate/revoke edge cases

## Test plan

- [x] All existing tests pass with hash-check enabled
- [x] JTI-1: Two tokens with same claims produce different hashes
- [x] JTI-2: After regenerate, old token rejected (401)
- [x] JTI-3: After revoke + re-issue, old token stays dead

### CODE REVIEW

**Review by wanguardd — CHANGES_REQUESTED** [review]
# Code Review - PR #44

## Overview

This PR implements a critical security enhancement for IC Token validation by adding runtime database hash-checking to prevent revoked and rotated tokens from remaining usable. The implementation introduces:

1. **JTI Claim (UUID v4)**: Guarantees token uniqueness even when issued in the same second
2. **Runtime Validation Method**: `validate_ic_token_runtime()` combining JWT verification with database hash-check
3. **Systematic Updates**: 6 runtime endpoints updated, 21 test files updated

**Verdict: NOT READY** - 12 findings identified (1 CRITICAL, 5 HIGH, 6 MEDIUM)

---

## Critical Issues (Must Fix Before Merge)

### C1. Fix Comment Missing Mandatory 3-Field Format

**Location:** `module/iron_control_api/src/ic_token.rs:238-242`

**Issue:** The `validate_ic_token_runtime` method includes fix documentation in doc comment format, but CLAUDE.md §2 requires inline 3-field comment format for all bug fixes.

**Current:**
```rust
/// Validate IC Token for runtime endpoints: JWT verification + hash-check against database
///
/// Fix(ic-token-invalidation): Runtime endpoints previously relied only on JWT signature
/// verification, allowing revoked or rotated tokens to remain usable. This method adds
/// a database hash-check to ensure only the current active token is accepted.
```

**Required:**
```rust
// Fix(ic-token-invalidation): Add runtime hash-check to prevent revoked tokens
// Root cause: Runtime endpoints only verified JWT signature, never checked if token was still active in database
// Pitfall: Never rely on JWT signature alone for stateful authentication - always validate against authoritative state
/// Validate IC Token for runtime endpoints: JWT verification + hash-check against database
pub async fn validate_ic_token_runtime(...) { }
```

**Authority:** CLAUDE.md §2 (Bug-Fixing Workflow), code_design.rulebook.md

---

## High Priority Issues (Should Fix Before Merge)

### H1. Timing Attack Vulnerability in Hash Comparison

**Location:** `module/iron_control_api/src/ic_token.rs:313`

**Issue:** Hash comparison uses standard string equality (`!=`), which is vulnerable to timing attacks allowing attackers to extract the stored hash character-by-character through response time measurements.

**Current:**
```rust
Some( Some( ref hash ) ) if *hash != token_hash =>
```

**Attack Vector:**
1. Attacker generates tokens with different hash values
2. Measures response time for each validation attempt
3. Identifies matching characters by timing differences
4. Reconstructs the stored hash over many requests

**Fix:**
```rust
// Add to Cargo.toml:
subtle = "2.5"

// Update ic_token.rs:313:
use subtle::ConstantTimeEq;

Some( Some( ref hash ) ) if hash.as_bytes().ct_eq(token_hash.as_bytes()).unwrap_u8() == 0 =>
```

**Why HIGH:** Requires many requests but directly exploitable. Combined with H2 (no rate limiting), this becomes more severe.

---

### H2. Missing Rate Limiting on Validation Endpoints

**Location:** All 6 updated endpoints

**Issue:** Token validation performs database queries on every request with no rate limiting. Budget endpoints explicitly disable rate limiting.

**Evidence:**
```rust
// src/routes/budget/state.rs:44-45
rate_limiting_enabled: false,
```

**Attack Vectors:**
1. **Brute-force timing attacks:** Unlimited validation requests to exploit H1
2. **Database DoS:** Flood validation endpoints to overload database
3. **Agent enumeration:** Try different agent_ids (mitigated by generic errors, but timing still leaks info)

**Recommendation:**
```rust
// Add per-IP rate limiting for runtime endpoints:
// - 100 requests/minute for token validation failures
// - 1000 requests/minute for successful validations
// - Exponential backoff after repeated failures
```

**Why HIGH:** Directly enables H1 timing attack exploitation and can cause service degradation.

---

### H3. IcTokenRuntimeError Lacks Display/Error Trait Implementation

**Location:** `module/iron_control_api/src/ic_token.rs:32-42`

**Issue:** The custom error enum doesnt implement `Display` or `std::error::Error` traits, making it inconsistent with Rust error handling conventions.

**Current:**
```rust
#[ derive( Debug ) ]
pub enum IcTokenRuntimeError
{
  InvalidToken( String ),
  InvalidAgentId( String ),
  TokenInactive( String ),
  DatabaseError( String ),
}
```

**Required:**
```rust
impl std::fmt::Display for IcTokenRuntimeError {
  fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
    match self {
      Self::InvalidToken(msg) => write!(f, "Invalid IC Token: {}", msg),
      Self::InvalidAgentId(msg) => write!(f, "Invalid agent_id: {}", msg),
      Self::TokenInactive(msg) => write!(f, "IC Token inactive: {}", msg),
      Self::DatabaseError(msg) => write!(f, "Database error: {}", msg),
    }
  }
}

impl std::error::Error for IcTokenRuntimeError {}
```

**Impact:** All 4 usage sites currently use `{:?}` debug formatting. This inconsistency makes error messages less user-friendly.

**Why HIGH:** Error handling consistency is crucial for maintainability.

---

### H4. Missing Database Error Tests

**Location:** Test coverage gap

**Issue:** No tests simulate database failures during `validate_ic_token_runtime()`.

**Risk:**
- Database errors during validation could return 500 errors with sensitive information
- Could fail open (accept invalid tokens)
- Could cause deadlocks in production

**Recommendation:**
```rust
#[ tokio::test ]
async fn test_validate_ic_token_database_error()
{
  // Create pool, then close it to simulate connection failure
  let pool = setup_test_db().await;
  pool.close().await;

  let manager = IcTokenManager::new( "secret".to_string() );
  let token = create_valid_token();

  // Validation should return DatabaseError
  let result = manager.validate_ic_token_runtime( &token, &pool ).await;
  assert!( matches!( result, Err( IcTokenRuntimeError::DatabaseError( .. ) ) ) );
}
```

**Why HIGH:** Production reliability - untested error paths are a risk.

---

### H5. Missing Concurrent Validation Tests

**Location:** Test coverage gap

**Issue:** No tests for race conditions during token regeneration while validation is in progress.

**Risk:** TOCTOU vulnerabilities - old token validated → regenerate happens → old token accepted.

**Recommendation:**
```rust
#[ tokio::test ]
async fn test_concurrent_validation_during_regenerate()
{
  let token_a = create_ic_token( &pool, agent_id, &manager ).await;

  // Start validation + regenerate concurrently
  let validation_task = tokio::spawn( validate_token_in_loop( token_a.clone() ) );
  let regenerate_task = tokio::spawn( regenerate_token( agent_id ) );

  // Both should complete without deadlock
  // Old token should eventually fail validation
}
```

**Why HIGH:** Concurrency bugs are hard to debug in production.

---

## Medium Priority Issues (Address Soon)

### M1. Inconsistent Error Handling Pattern Across Endpoints

**Location:** `agent_provider_key.rs`, `analytics/ingestion.rs`, `budget/handshake.rs`, `budget/refresh.rs`

**Issue:** All four endpoints handle `IcTokenRuntimeError` identically, but the code is duplicated 4 times (45-47 lines each). Violates Anti-Duplication Principle.

**Recommendation:**
Extract to helper function in `ic_token.rs`:
```rust
pub async fn validate_ic_token_for_endpoint(
  manager: &IcTokenManager,
  pool: &SqlitePool,
  token: &str,
) -> Result<(i64, IcTokenClaims), impl IntoResponse> { /* ... */ }
```

---

### M2. sha256_hash Function Lacks Unit Tests

**Location:** `module/iron_control_api/src/ic_token.rs:22-28`

**Issue:** Public utility function used across 6 files has no dedicated unit tests.

**Recommendation:**
```rust
#[cfg(test)]
mod tests {
  use super::*;

  #[test]
  fn test_sha256_hash_deterministic() {
    let token = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.test";
    let hash1 = sha256_hash(token);
    let hash2 = sha256_hash(token);
    assert_eq!(hash1, hash2);
  }

  #[test]
  fn test_sha256_hash_format() {
    let hash = sha256_hash("test");
    assert_eq!(hash.len(), 64); // 64 hex chars
    assert!(hash.chars().all(|c| c.is_ascii_hexdigit()));
  }
}
```

---

### M3. No Transaction Isolation for Token Regeneration

**Location:** Token regeneration flow (inferred)

**Issue:** Token regeneration performs sequential operations without transaction isolation:
1. Check if agent exists
2. Generate new token
3. Update hash in database

**Race Condition:** Two concurrent regenerate requests → both generate tokens → second overwrites first → first token holder has invalid token.

**Recommendation:**
```rust
let mut tx = pool.begin().await?;
let ic_token = generate_token()?;
sqlx::query("UPDATE agents SET ic_token_hash = ? WHERE id = ? FOR UPDATE")
    .bind(&token_hash)
    .bind(agent_id)
    .execute(&mut *tx)
    .await?;
tx.commit().await?;
```

---

### M4. Generic Error Messages May Leak Timing Information

**Location:** All validation endpoints

**Issue:** Error type branching may leak information via timing:
- DatabaseError path logs with `tracing::error!` (slower)
- Other errors log with `tracing::warn!` (faster)

**Recommendation:**
```rust
Err(e) => {
    tracing::warn!("IC Token validation failed: {:?}", e);
    // Add small constant delay to normalize timing (1-5ms)
    tokio::time::sleep(Duration::from_millis(2)).await;
    return UNAUTHORIZED;
}
```

---

### M5. Analytics Endpoint Missing Explicit Tests

**Location:** `/api/v1/analytics/event`

**Issue:** Endpoint updated to use runtime validation but lacks explicit test coverage for rotated token rejection.

**Recommendation:**
```rust
#[ tokio::test ]
async fn test_analytics_ingestion_with_rotated_token() {
  // Similar to test_regenerated_token_invalidates_old_token_via_hash_check
  // but for analytics endpoint
}
```

---

### M6. Missing Unit Tests for validate_ic_token_runtime()

**Location:** Test coverage gap

**Issue:** Only end-to-end tests exist. No direct unit tests for the functions error branches.

**Recommendation:** Add unit test file testing each error branch in isolation.

---

## Positive Observations

### Architecture & Organizational Excellence

✅ **Anti-Duplication Principle - sha256_hash Consolidation**
- Before: `sha256_hash` duplicated in `routes/ic_token.rs`
- After: Made public in `ic_token.rs`, removed duplicate
- Single source of truth for token hashing

✅ **No Backup Files Created**
- Zero files matching `*_backup`, `*_old`, `*_v[0-9]`, `*_legacy`, `*.bak`
- All changes are clean replacements

✅ **No Git Attribution Footers**
- Commit messages are clean and professional
- No "Generated with Claude Code" footers
- Compliant with CLAUDE.md §1 requirements

✅ **Hierarchical Separation Maintained**
- Production code in `src/ic_token.rs`
- Runtime endpoint changes in `src/routes/`
- Test infrastructure in `tests/common/`
- No cross-boundary violations

---

### Security Excellence

✅ **JTI Claim for Token Uniqueness**
- Guarantees unique JWT signatures even for same-second generation
- Uses UUID v4 (122 bits entropy)
- Well-documented with test coverage

✅ **Hash-Based Token Validation**
- Only SHA-256 hash stored in database (never plaintext)
- Defense in depth: both JWT and hash must be valid

✅ **Parameterized SQL Queries**
- All queries use parameter binding
- Zero SQL injection risk

✅ **Input Validation on agent_id**
- Format validation (prefix check)
- Type validation (numeric parse)
- Range validation (positive check)

✅ **Agent Enumeration Protection**
- Both "agent not found" and "token revoked" return same error
- Prevents attackers from discovering valid agent IDs

---

### Testing Excellence

✅ **ADR-007 Compliance - No Mocking**
- All tests use real SQLite databases (in-memory)
- Real IC Token generation/validation
- Real cryptographic operations
- Real HTTP routers

✅ **Strong Test Isolation**
- Each test gets independent database instance
- TempDir Drop pattern for automatic cleanup
- Tests can run in parallel

✅ **Comprehensive Security Test Scenarios**
- JTI uniqueness (same-second generation)
- Token regeneration invalidation (end-to-end)
- Token revoke-then-reissue (old token stays dead)
- Expired token rejection

✅ **Excellent Assert Messages**
- 95%+ assert coverage with descriptive messages
- "LOUD FAILURE" pattern throughout

---

### Code Quality Excellence

✅ **Perfect Custom Codestyle Compliance**
- All code uses proper 2-space indentation
- No cargo fmt patterns detected
- Consistent with project custom style

✅ **Clear Test Structure**
- Arrange-Act-Assert pattern
- Well-documented scenarios
- Descriptive test names

✅ **Comprehensive Documentation**
- Extensive doc comments for new functions
- Clear validation step enumeration
- Good use of markdown formatting

---

## Summary

**Code Implementation:** A (Excellent)
**Test Coverage:** A- (Strong with minor gaps)
**Security Engineering:** B+ (Good with timing attack fix needed)
**Documentation:** B (Needs critical format fixes)

### Merge Readiness Checklist

- [x] **C1** - Fix comment in 3-field format
- [x] **H1** - Fix timing attack vulnerability (constant-time comparison)
- [x] **H2** - Add rate limiting to validation endpoints
- [x] **H3** - Implement Display/Error traits for IcTokenRuntimeError
- [x] **H4** - Add database error tests
- [x] **H5** - Add concurrent validation tests

Once these 6 items are addressed, this PR is ready to merge.

### Post-Merge Work (Next Sprint)

- [ ] **M1-M6** - Address medium priority issues
- [ ] Run coverage analysis and fill gaps
- [ ] Monitor production metrics after deploy

---

## Conclusion

PR 44 is a **well-engineered security enhancement** with strong architectural discipline and comprehensive test coverage. The core implementation is sound and follows project standards. The identified issues are primarily documentation format violations and security hardening opportunities that, while important, do not undermine the fundamental correctness of the implementation.

**Key Strengths:**
- Solves critical security vulnerability (token invalidation)
- Strong test coverage with real implementations
- Perfect custom codestyle compliance
- Excellent adherence to organizational principles
- Comprehensive security test scenarios

**Key Weaknesses:**
- Documentation format violation (critical but easily fixed)
- Timing attack vulnerability (high priority fix)
- Missing rate limiting (high priority enhancement)
- Test coverage gaps for error scenarios

**Recommendation:** Address the 6 critical/high-priority issues, then merge with confidence. The remaining issues can be addressed in follow-up PRs without blocking this security fix.

File: module/iron_control_api/src/ic_token.rs [hunk 1]
```diff
@@ -11,48 +11,95 @@
...
+use std::time::{Duration, Instant, SystemTime, UNIX_EPOCH};
+use subtle::ConstantTimeEq;
+use uuid::Uuid;
+
+/// Compute SHA-256 hash of a token string (hex-encoded)
+pub fn sha256_hash(token: &str) -> String {
+  let mut hasher = Sha256::new();
+  hasher.update(token.as_bytes());
+  format!("{:x}", hasher.finalize())
+}
```
Comment (tihilya) [inline module/iron_control_api/src/ic_token.rs]: pub fn sha256_hash(token: &str) -> String {
format!("{:x}", Sha256::digest(token.as_bytes()))
}

**Review by wanguardd — CHANGES_REQUESTED** [review]
## PR Review: IC Token Runtime Validation with Hash-Check and JTI Uniqueness

**Outcome**: NOT READY - 1 critical security issue requires immediate fix

**Findings Summary**:
- 🔴 CRITICAL: 1 (blocking)
- 🟡 HIGH: 1 (documentation)
- 🟢 MEDIUM: 1 (code standards)

---

## 🔴 CRITICAL Security Issue (BLOCKING)

### [C-1] `#[derive(Debug)]` on `IcTokenManager` exposes HMAC signing secret

**Location**: `module/iron_control_api/src/ic_token.rs:190`

**Evidence**:
```rust
/// IC Token manager for generating and validating IC Tokens
#[derive(Debug)]
pub struct IcTokenManager {
  secret: String,
}
```

**Issue**: The auto-derived `Debug` impl exposes the HMAC signing secret in logs, error messages, and debug output. Any code that logs or formats `IcTokenManager` with `{:?}` will leak the secret.

**Impact**: CRITICAL - Leaked HMAC secret allows forging valid IC Tokens, bypassing all authentication.

**Fix Required**:
```rust
impl core::fmt::Debug for IcTokenManager {
  fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
    f.debug_struct("IcTokenManager")
      .field("secret", &"<redacted>")
      .finish()
  }
}
```

**Validation Command**:
```bash
$ rg '#\[derive\(Debug\)\]' -A 2 module/iron_control_api/src/ic_token.rs | grep -A 2 'IcTokenManager'
```

---

## 🟡 HIGH Priority Issues

### [H-1] Protocol 005 documentation not updated with new security mechanisms

**Evidence**: The new security features are not documented in:
- `docs/protocol/005_budget_control_protocol.md` - no mention of jti, hash-check, rate limiting, timing padding
- `spec/requirements.md` - no mention of IC Token runtime validation security

**Missing Documentation**:
1. `jti` (JWT ID) claim - UUID v4 for guaranteed token uniqueness
2. Runtime hash-check validation against database
3. Unified 401 "Invalid IC Token" response (enumeration prevention)
4. Per-token rate limiting (20 failures / 60-second window)
5. 50ms timing padding (VALIDATION_TIMING_TARGET)

**Recommendation**: Add security section to Protocol 005 documenting:
- [ ] Hash-check validation flow
- [ ] JTI claim purpose and format
- [ ] Rate limiting configuration
- [ ] Timing attack mitigation

---

## 🟢 MEDIUM Priority Issues

### [M-1] JTI security tests missing `bug_reproducer(issue-NNN)` markers

**Location**: `module/iron_control_api/tests/budget_security_ic_token.rs`

**Tests Affected**:
- `test_jti_guarantees_unique_token_hashes` (line 281)
- `test_regenerated_token_invalidates_old_token_via_hash_check` (line 342)
- `test_revoke_then_reissue_old_token_stays_dead` (line 470)

**Project Standard**: Bug-fixing tests should include `// test_kind: bug_reproducer(issue-NNN)` comment per `code_design.rulebook.md`.

**Evidence**: Tests document bug fixes but lack traceable issue markers:
```rust
/// # Risk
/// CRITICAL - This is the core scenario that Task 001 fixes
#[tokio::test]
async fn test_regenerated_token_invalidates_old_token_via_hash_check() {
```

**Recommendation**: Add `// test_kind: bug_reproducer(issue-001)` or similar markers.

---

## ✅ Positive Observations

1. **Code compiles and tests pass**: All 50 tests in `budget_security_ic_token` pass
2. **Clippy clean**: No warnings with `-D warnings`
3. **Security design is sound**:
   - Constant-time hash comparison with `subtle::ConstantTimeEq`
   - Timing padding prevents timing oracle attacks
   - Rate limiting mitigates brute force
   - Unified error response prevents enumeration
4. **JTI implementation correct**: UUID v4 guarantees unique tokens

---

## Action Items

- [ ] **BLOCKING**: Replace `#[derive(Debug)]` with manual impl that redacts `secret`
- [ ] Update Protocol 005 documentation with security mechanisms
- [ ] Add `bug_reproducer` markers to JTI tests

---

**Review Validation**:
- Branch: `fix/runtime-token-validation`
- Commit: `2966f67dfe8e102dbedc856df951cd106a2a3df2`
- Tests: `cargo test --package iron_control_api --test budget_security_ic_token` ✅
- Clippy: `cargo clippy --package iron_control_api -- -D warnings` ✅
